### PR TITLE
Feature/9179 custom fields enhancement

### PR DESCRIPTION
**Pull request by Anmol79**
http://pm.7vals.com/issues/9179
Server side changes:
https://github.com/7Vals/ezofficeinventory/pull/8613/files

### CODE REVIEW

**Comment by Anmol79** [comment]
Cycle 0

File: app/models/custom_attributes/custom_attribute_value.rb [hunk 1]
```diff
@@ -9,12 +9,17 @@ def name
 
       def apply_default_value
         unless custom_attribute_defn.default_value.blank?
-          self.value = custom_attribute_defn.default_value
+          if custom_attribute_defn.dropdown_type?
```
Comment (Anmol79) [inline app/models/custom_attributes/custom_attribute_value.rb]: create another method has_options? and use that

File: app/models/custom_attributes/custom_attribute_value.rb [hunk 2]
```diff
@@ -9,12 +9,17 @@ def name
...
+            custom_attribute_option  = custom_attribute_defn.custom_attribute_options.find_by id: custom_attribute_defn.default_value
```
Comment (Anmol79) [inline app/models/custom_attributes/custom_attribute_value.rb]: refactor and validate default value on deletion

File: app/models/custom_attributes/custom_attribute_value.rb [hunk 3]
```diff
@@ -9,12 +9,17 @@ def name
...
+            self.value               = custom_attribute_option.try(:id)
+          else
+            self.value = custom_attribute_defn.default_value
+          end
         end
       end
 
       def value_present?
-        string_value.present? || double_value.present? || integer_value.present? || !boolean_value.nil?
+        string_value.present? || double_value.present? || integer_value.present? || !boolean_value.nil? || custom_attribute_option_values.any?
```
Comment (Anmol79) [inline app/models/custom_attributes/custom_attribute_value.rb]: mandatory check for po

File: app/views/custom_attributes/custom_attribute_definitions/_custom_attribute_option_fields.html.erb [hunk 4]
```diff
@@ -0,0 +1,19 @@
+<% option_deletion_alert_url ||= nil %>
```
Comment (Anmol79) [inline app/views/custom_attributes/custom_attribute_definitions/_custom_attribute_option_fields.html.erb]: confirm_option_deletion_url

File: lib/generators/custom_attributes/initialize_generator.rb [hunk 5]
```diff
@@ -41,12 +43,17 @@ def create_derived_custom_attribute_values
...
         create_custom_attribute_value_sub_class("DateTime")
         create_custom_attribute_value_sub_class("Boolean")
+        create_custom_attribute_value_sub_class("Dropdown")
       end
 
       def copy_inline_editing_js_file
         template "custom_attributes_inline_editing.js.erb", "public/javascripts/vendors/custom_attributes/inline_editing.js"
       end
 
+      def copy_dropdown_js_file
```
Comment (Anmol79) [inline lib/generators/custom_attributes/initialize_generator.rb]: try if it gives option on generation

File: lib/generators/custom_attributes/templates/custom_attribute_option_value_model.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+  include CustomAttributes::CustomAttributeOptionValue
+  belongs_to :custom_attribute_option, class_name: "<%= name %>CustomAttributeOption", foreign_key: "<%= singular_name %>_custom_attribute_option_id"
+  belongs_to :custom_attribute_value, class_name: "<%= name %>CustomAttributeValue", foreign_key: "<%= singular_name %>_custom_attribute_value_id"
+  has_one    :<%= singular_name %>, through: :<%= singular_name %>_custom_attribute_value
+  has_one    :custom_attribute_defn, through: :custom_attribute_value
+
+  <% if options.tenant %>
+    belongs_to :<%= options.tenant %>_id
+    scope :current_tenant, -> <%= options.tenant %>_id { where(<%= options.tenant %>_id: <%= options.tenant %>_id) }
+    # Uncomment the following if you want to define a default_scope instead
```
Comment (Anmol79) [inline lib/generators/custom_attributes/templates/custom_attribute_option_value_model.rb]: remove

File: lib/generators/custom_attributes/templates/custom_attributes_dropdown_option.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  $('.custom_attribute_options_content').append(content.replace(regexp, new_id));
+  $('.dropdown-ca-default-select').append("<option></option>");
+}
+
+function remove_option_from_dropdown_field(link){
+  var element = $(link).closest(".custom_attribute_options_content");
+  var option  = $(link).closest(".custom_dropdown_option");
+  if (element.find(".custom_dropdown_option:visible").length == 1){
+    alert("You have to have atleast one option");
+    return;
```
Comment (Anmol79) [inline lib/generators/custom_attributes/templates/custom_attributes_dropdown_option.js]: false

File: app/models/custom_attributes/custom_attribute_value.rb (hunk 3 above)
Comment (Anmol79) [reply app/models/custom_attributes/custom_attribute_value.rb]: done

File: lib/generators/custom_attributes/initialize_generator.rb (hunk 5 above)
Comment (Anmol79) [reply lib/generators/custom_attributes/initialize_generator.rb]: tested

File: app/models/custom_attributes/custom_attribute.rb [hunk 8]
```diff
@@ -5,6 +5,7 @@ class CustomAttribute
     TYPE_DECIMAL        = 'decimal'
     TYPE_TEXT           = 'text'
     TYPE_BOOLEAN        = 'boolean'
+    TYPE_DROPDOWN       = 'dropdown'
```
Comment (Zeeshan-Siddhu) [inline app/models/custom_attributes/custom_attribute.rb]: can we remove this file in future

File: app/views/custom_attributes/custom_attribute_definitions/_form_fields_only.html.erb [hunk 9]
```diff
@@ -18,12 +19,30 @@
...
-<% unless f.object.try(:boolean_type?) %>
+<div class="clearfix"></div>
+<div class="custom-attribute-definition-options <%= 'hide' if f.object.blank? || !f.object.has_options? %>">
+  <div class="custom_attribute_options_content sort-custom-attribute-options" data-update-url="<%= position_update_url %>">
+    <%= f.fields_for :custom_attribute_options do |g| %>
+      <%= render partial: 'custom_attribute_option_fields', locals: { f: g, confirm_option_deletion_url: confirm_option_deletion_url } %>
+    <% end %>
+  </div>
+  <% 
+    new_object = f.object.class.reflect_on_association(:custom_attribute_options).klass.new
```
Comment (Zeeshan-Siddhu) [inline app/views/custom_attributes/custom_attribute_definitions/_form_fields_only.html.erb]: discuss in final review/ saqib

File: app/views/custom_attributes/custom_attribute_definitions/_form_fields_only.html.erb [hunk 10]
```diff
@@ -18,12 +19,30 @@
...
+    fields     = f.fields_for(:custom_attribute_options, new_object, child_index: "new_custom_attribute_options") do |builder|
+      render("custom_attribute_option_fields", f: builder)
+    end
+  %>
+  <%= link_to_function('Add Options', "addOptionToDropdownField('custom_attribute_options', '#{escape_javascript(fields)}')") %>
```
Comment (Zeeshan-Siddhu) [inline app/views/custom_attributes/custom_attribute_definitions/_form_fields_only.html.erb]: space

**Comment by Zeeshan-Siddhu** [comment]
Cycle 2

File: app/models/custom_attributes/custom_attribute_option_value.rb [hunk 11]
```diff
@@ -0,0 +1,10 @@
+module CustomAttributes
```
Comment (saqibbukhari) [inline app/models/custom_attributes/custom_attribute_option_value.rb]: remove

File: app/models/custom_attributes/dropdown_value.rb [hunk 12]
```diff
@@ -0,0 +1,25 @@
+module CustomAttributes
+  module DropdownValue
+    def value=(custom_attribute_option_id)
+      custom_attribute_option = custom_attribute_options.find_by id: custom_attribute_option_id
+      if (custom_attribute_option_value = custom_attribute_option_values.first).present?
+        if custom_attribute_option.present?
+          custom_attribute_option_value.custom_attribute_option = custom_attribute_option
+          custom_attribute_option_value.save
+        else
+          custom_attribute_option_value.destroy
```
Comment (saqibbukhari) [inline app/models/custom_attributes/dropdown_value.rb]: redmine for mandatory check and add it in December

File: app/views/custom_attributes/custom_attribute_definitions/_custom_attribute_option_fields.html.erb [hunk 13]
```diff
@@ -0,0 +1,19 @@
+<% confirm_option_deletion_url ||= nil %>
```
Comment (saqibbukhari) [inline app/views/custom_attributes/custom_attribute_definitions/_custom_attribute_option_fields.html.erb]: test and check for local assign

File: app/views/custom_attributes/dropdown_values/_edit_dropdown_value.html.erb [hunk 14]
```diff
@@ -0,0 +1,32 @@
+<% form       ||= nil %>
+<% cssClasses ||= {} %>
+<% required   ||= false %>
+<% if form %>
+  <% # BEGIN Field part of form %>
+    <%= form.fields_for :custom_attributes do |custom_fields| %>
+      <% label_text = dropdown_value.name %>
+      <% label_text += content_tag(:span, '*', class: cssClasses[:required]) if required %>
+      <%= custom_fields.label dropdown_value.name.to_sym, label_text.html_safe, class: cssClasses[:title] %>
```
Comment (saqibbukhari) [inline app/views/custom_attributes/dropdown_values/_edit_dropdown_value.html.erb]: sanitize

File: lib/generators/custom_attributes/templates/migration.rb [hunk 15]
```diff
@@ -39,5 +39,27 @@ def change
     add_foreign_key :<%= singular_table_name %>_custom_attribute_values, :<%= options.tenant.pluralize %>, column: :<%= options.tenant.underscore %>_id, on_delete: :cascade
   <% end %>
 
+    create_table (:<%= singular_table_name %>_custom_attribute_options) do |t|
+      t.text :label, index: { length: 767 }
+      t.integer :position
```
Comment (saqibbukhari) [inline lib/generators/custom_attributes/templates/migration.rb]: index

File: app/models/custom_attributes/dropdown_value.rb (hunk 12 above)
Comment (Anmol79) [reply app/models/custom_attributes/dropdown_value.rb]: http://pm.7vals.com/issues/9485

File: app/views/custom_attributes/custom_attribute_definitions/_custom_attribute_option_fields.html.erb (hunk 13 above)
Comment (Anmol79) [reply app/views/custom_attributes/custom_attribute_definitions/_custom_attribute_option_fields.html.erb]: tested

**Pull request merged**